src/test/java/org/springframework/data/elasticsearch/junit/jupiter/SpringDataElasticsearchExtension.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.junit.jupiter;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.platform.commons.support.AnnotationSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This extension class check in the {@link #beforeAll(ExtensionContext)} call if there is already a Elasticsearch
+ * cluster connection defined in root store. If no, the connection to the cluster is defined according to the
+ * configuration, starting a local node if necessary. The connection is stored and will be closed when the store is
+ * shutdown at the end of all tests.
+ *
+ * @author Peter-Josef Meisch
+ */
+public class SpringDataElasticsearchExtension implements BeforeAllCallback {
+
+	public static final String SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL = "SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL";
+	private static final Logger LOGGER = LoggerFactory.getLogger(SpringDataElasticsearchExtension.class);
+
+	private static final String STORE_KEY = ClusterConnection.class.getSimpleName();
+
+	private static final Lock initLock = new ReentrantLock();
+
+	@Override
+	public void beforeAll(ExtensionContext extensionContext) throws Exception {
+		LOGGER.debug("initializing");
+
+		ClusterConnection clusterConnection = getClusterConnection(extensionContext);
+		Class<?> testClass = extensionContext.getRequiredTestClass();
+
+		AnnotationSupport.findAnnotatedFields(testClass, SpringDataElasticsearch.class).forEach(field -> {
```

Comment:
Probably we should use only the built-in injection mechanism instead of assigning values to `static` fields.

Reply:
This runs before the spring context is set up. When Spring is set up, this information must already be available to be able to inject it into the spring beans

Reply:
Yeah, therefore it would make sense to associate the connection details with the Spring context. I'm not sure how much customization we can apply for `SpringExtension`. I think we would need to call `ConfigurableListableBeanFactory.registerResolvableDependency(…)` at some point. Maybe @sbrannen can hint us in the right direction.

Reply:
I was thinking about extending `SpringExtension` but am not sure if this would be the right way. So I'd be glad for an hint how to get this logic of setting up the ClusterConnection only once for the test run and shutitng it down at the end of the tests coupled to Spring injection to have this info as a bean available.

Reply:
Basically, we need to bridge resources provided by our Spring Data ES Extension into the Spring context so Spring beans can utilize it. We're on the same page that the connection details lifecycle is managed by our Spring Data ES Extension.

Reply:
> Yeah, therefore it would make sense to associate the connection details with the Spring context. I'm not sure how much customization we can apply for `SpringExtension`. I think we would need to call `ConfigurableListableBeanFactory.registerResolvableDependency(…)` at some point. Maybe @sbrannen can hint us in the right direction.

If that's what you want to do, you could implement that in a `TestExecutionListener` (Spring API, not JUnit Platform) like I demonstrated here: https://github.com/spring-projects/spring-framework/issues/12947.

Reply:
You might also want to consider implementing a custom `org.springframework.test.context.ContextCustomizer`.

Spring Boot's testing support relies heavily on the `ContextCustomizer` SPI for many of its features.

Reply:
Thanks a lot, Sam. These are the APIs we were looking for.

Reply:
thanks for all this information; I'll have a deeper look a t this the next week